### PR TITLE
More cppcheck fixes

### PR DESCRIPTION
**Pull request by cujomalainey**

### CODE REVIEW

**Review by marc-hb — DISMISSED** [review]

**Comment by marc-hb** [comment]
> I can take a look, but it might not work given that this needs visibility of all headers and such to be able to infer properly.

Neither the sof-test script nor the github action I gave "hide" anything. They only extract a list of modified files that can be passed as arguments to cppcheck. That does not stop cppcheck from looking at any other header or other file it needs.

PS: any idea why I get only the `init.c` warning on the files fixed by this PR and not the other warnings? I'm using the same options with cppcheck 1.90

**Comment by marc-hb** [comment]
> which is why I have this only turned on for the most serious of errors currently, to reduce the noise.

In fact checking only the files modified in the PR allows turning on many more checks.

**Comment by cujomalainey** [comment]
> > which is why I have this only turned on for the most serious of errors currently, to reduce the noise.
> 
> In fact checking only the files modified in the PR allows turning on many more checks.
> 
> Running fewer checks on ALL files is best done as a nightly build, not as a PR check.

So errors on push for all files, all for only files touched by the PR on PR?

**Comment by cujomalainey** [comment]
> > > which is why I have this only turned on for the most serious of errors currently, to reduce the noise.
> > 
> > 
> > In fact checking only the files modified in the PR allows turning on many more checks.
> > Running fewer checks on ALL files is best done as a nightly build, not as a PR check.
> 
> So errors on push for all files, all for only files touched by the PR on PR?

For reference I will start with error on the second, there are simply way too many warning/info/style issues to open up that fire hose right now

**Comment by cujomalainey** [comment]
So it turns out cppcheck is really out of date in github actions which is yielding some bogus results, i'll see what I can do

**Comment by cujomalainey** [comment]
I have punted the CI commit to my [cppcheck-ci-enable](https://github.com/cujomalainey/sof/tree/cppcheck-ci-enable) branch as ubuntu 20.04 seems to be far behind running pre cppcheck 2.0. I don't think it is wise to enable this in CI until at least 2.3. Turning this PR into a generic fixes PR in the mean time.

**Comment by cujomalainey** [comment]
Also to all the die hard C fans here, I apologize for my sass in my commits against C, I have been learning rust in my spare time and am quickly falling in love.

**Comment by marc-hb** [comment]
> I have punted the CI commit to my cppcheck-ci-enable branch as ubuntu 20.04 seems to be far behind running pre cppcheck 2.0. 

I found no Github plan to provide a more recent Ubuntu  version any  time soon https://github.com/actions/virtual-environments

The recommendation is to use Docker and that's what we already do for other jobs. Did you consider that?

> I apologize for my sass in my commits against C,

I was looking forward to some cheap shots but didn't find any :-(

**Comment by cujomalainey** [comment]
> > I have punted the CI commit to my cppcheck-ci-enable branch as ubuntu 20.04 seems to be far behind running pre cppcheck 2.0.
> 
> I found no Github plan to provide a more recent Ubuntu version any time soon https://github.com/actions/virtual-environments
> 
> The recommendation is to use Docker and that's what we already do for other jobs. Did you consider that?

Nope, I will look into that

> 
> > I apologize for my sass in my commits against C,
> 
> I was looking forward to some cheap shots but didn't find any :-(

See  the crossover commit

**Comment by marc-hb** [comment]
> See the crossover commit

Oops, missed it.

Integer promotion rules give me a headache every time I have to look them up.

For even more fun try this:
```diff
--- a/src/arch/xtensa/CMakeLists.txt
+++ b/src/arch/xtensa/CMakeLists.txt
@@ -124,7 +124,7 @@ target_compile_options(sof_options INTERFACE ${stdlib_flag} -fno-inline-function
 target_compile_options(sof_options INTERFACE
 	$<$<COMPILE_LANGUAGE:C>:
 		-${optimization_flag} -g
-		-Wall -Werror
+		-Wall -Wsign-conversion -Wsign-compare
 		-Wl,-EL
 		-Wmissing-prototypes
 		-Wpointer-arith
```

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Just one doxygen item to fix
```
[0/1] cd /home/runner/work/sof/sof/docbuild && doxygen sof.doxygen
/home/runner/work/sof/sof/src/include/sof/ipc/common.h:183: error: argument 'hdr' of command @param is not found in the argument list of ipc_cmd(ipc_cmd_hdr *_hdr) (warning treated as error, aborting now)
FAILED: CMakeFiles/doc /home/runner/work/sof/sof/docbuild/CMakeFiles/doc 
cd /home/runner/work/sof/sof/docbuild && doxygen sof.doxygen
```

File: src/include/sof/ipc/common.h
Comment (marc-hb) [inline src/include/sof/ipc/common.h:190]: Just curious what is the warning here.

Comment (cujomalainey) [reply src/include/sof/ipc/common.h:190]: That the signature in the header didnt match the implementation signature in the c files

Comment (cujomalainey) [reply src/include/sof/ipc/common.h:190]: See https://github.com/thesofproject/sof/blob/main/src/ipc/ipc3/handler.c#L1513

Comment (marc-hb) [reply src/include/sof/ipc/common.h:190]: I disagree 100% with this check. I don't really care about this particular line but I think this check should be turned off.

The function signature is the same otherwise the code would not compile. That's how the compiler reports an actual signature mismatch:
```
sof/src/ipc/ipc3/handler.c:1513:6: error: conflicting types for 'ipc_cmd'
 1513 | void ipc_cmd(ipc_cmd_hdr *_hdr)
      |      ^~~~~~~
src/include/sof/ipc/common.h:191:6: note: previous declaration of 'ipc_cmd' was here
  191 | void ipc_cmd(ipc_cmd_hdr **hdr);
      |      ^~~~~~~
```
Parameter names in function declarations are totally optional and purely informative.

Having slightly different parameter names between the declaration and the definition is a very useful C feature. For instance:
- Slightly longer and more descriptive parameter names in the function declaration because there is no context.
- Slightly shorter parameter names in the function definition because there is all the context and repeating  long symbol names over and over in the function body makes  the code more verbose and less readable.

In this particular case neither `_hdr` nor `hdr` adds any information. The difference between the two serves a different purpose here: the underscore `_` has very specific and fairly clear meaning in the function code that makes no sense in the function declaration.

Again I don't really care about this particular line but it's a good example of how counter-productive this check is.

Comment (cujomalainey) [reply src/include/sof/ipc/common.h:190]: This was a style level violation, the downside is some cppcheck checks that are low level informative/style are actually reporting serious issues such as memory leaks. We can disable this specific check, but style as a general thing should be on kept I think.

Comment (marc-hb) [reply src/include/sof/ipc/common.h:190]: > We can disable this specific check, but style as a general thing should be on kept I think.

I don't know about cppcheck in general, my comment was laser-focused on this particular violation.

Comment (lgirdwood) [reply src/include/sof/ipc/common.h:190]: Lets fix the doxygen comment first and merge the fixes, we can continue cppcheck discussion afterwards.

Comment (cujomalainey) [reply src/include/sof/ipc/common.h:190]: @marc-hb by "check" I mean this type of style violation. I.e. invoke cppcheck with --suppress="this_type_of_check"

@lgirdwood I will do it next week, I am OOO friday-monday

File: src/audio/crossover/crossover.c
Comment (paulstelian97) [inline src/audio/crossover/crossover.c:97]: Someone needs to check that all calls to this are error-checked now, hopefully there's no hidden error but just in case.

**Review by paulstelian97 — APPROVED** [review]
Makes sense, good enough for me.

File: src/include/sof/ipc/common.h
Comment (plbossart) [reply src/include/sof/ipc/common.h:190]: > I disagree 100% with this check. I don't care about this particular line but I think this check should be turned off.
> 
> The function signature is the same otherwise the code would not compile. That's how the compiler reports an actual signature mismatch:
> 
> ```
> sof/src/ipc/ipc3/handler.c:1513:6: error: conflicting types for 'ipc_cmd'
>  1513 | void ipc_cmd(ipc_cmd_hdr *_hdr)
>       |      ^~~~~~~
> src/include/sof/ipc/common.h:191:6: note: previous declaration of 'ipc_cmd' was here
>   191 | void ipc_cmd(ipc_cmd_hdr **hdr);
>       |      ^~~~~~~
> ```
> 
> Parameter names in function declarations are totally optional and purely informative.
> 
> Having slightly different parameter names between the declaration and the definition is a very useful C feature. For instance:
> 
> * Slightly longer and more descriptive parameter names in the function declaration because there is no context.
> * Slightly shorter parameter names in the function definition because there is all the context and repeating  long symbol names over and over in the function body makes  the code more verbose and less readable.
> 
> In this particular case neither `_hdr` nor `hdr` adds any information. The difference between the two serves a different purpose here: the underscore `_` has a very specific and fairly clear meaning in the function code that makes no sense in the function declaration.
> 
> Again I don't care about this particular line but it's a good example of how counter-productive this check is.

No. Let's not be academic and rely on compilers.

it's a good practice to force variables to have the same names in .h and .c files. Usually this check raises more issues. We've cleaned this up for the kernel as well and the findings were priceless.

**Review by plbossart — APPROVED** [review]
static analysis is always right.

File: src/include/sof/ipc/common.h
Comment (marc-hb) [reply src/include/sof/ipc/common.h:190]: > No. Let's not be academic

I'm the one advocating for more flexibility and fewer rules in this particular case.

> and rely on compilers.

You mean the C standard. Parameters in declarations are optional and useless for and ignored by _all_ compilers.

> the findings were priceless.

One example?

Comment (plbossart) [reply src/include/sof/ipc/common.h:190]: > You mean the C standard. Parameters in declarations are optional and useless for and ignored by _all_ compilers.

Let's put it differently: Why on earth would a developer make a conscious attempt to name things differently in the header and the implementation? It's permitted by the standard but it's really stupid.

if you want examples, you can take a look at the cleanups I did for the kernel with 100+ of such changes. Inspecting the code and comparing declaration/definition helped flag additional issues with redundant inits and NULL pointer dereferences. It's really a telltale sign of code that needs to be reworked. Don't dismiss it on the grounds that the standard allows for it.

Comment (marc-hb) [reply src/include/sof/ipc/common.h:190]: > Why on earth would a developer make a conscious attempt to name things differently in the header and the implementation? It's permitted by the standard but it's really stupid.

I gave two examples above.

>  look at the cleanups I did for the kernel with 100+ of such changes.

I will try to find a few. How long ago?

> flag additional issues with redundant inits and NULL pointer dereferences

For now I really can't imagine how something ignored by all compilers can cause problems that bad.

> static analysis is always right.

I wish.

Comment (plbossart) [reply src/include/sof/ipc/common.h:190]: Did you read my point: it's a sign that the code developers did not maintain consistency between different parts of the code, and a likely sign of more errors. So when we see such problems, we also look at the rest of the function. it's the same with initialized variables or duplicate initialization, usually that means error handling is not done right.

I don't see what's objectionable about this. It's not like the SOF code is perfect or even good.

Another example: it's perfectly legal to have shadowed variables *because the compiler knows what to do*. And yet this is a real problem leading to bugs.

Comment (marc-hb) [reply src/include/sof/ipc/common.h:190]: > Did you read my point: it's a sign that the code developers did not maintain consistency between different parts of the code, and a likely sign of more errors. So when we see such problems, we also look at the rest of the function.

I did miss that indirection sorry. So this is like looking for typos and then performing a second code review pass where the typo was found. I'm not surprised at all that a second pass of code review finds bugs considering how rushed the first pass can be but is this really the most efficient to find problems and does this find the most important issues? How about - for instance - spending crucially limited time and expertise raising the bar on the first code review pass instead?

There is a gazillion of analyzers and checks that go beyond compilers to try and correct C limitations and design issues. Turning them all at once usually drowns important issues in an ocean of minor or irrelevant issues or false positives. I don't even care so much about that particular check actually but I care a lot about finding the right balance of cppcheck settings so that @cujomalainey can finally enable it in CI, something he hasn't been able to do  because of the aforementioned "ocean" as he stated above, see Aug. 26th comments. For evidence that I care a 5 seconds look at https://github.com/thesofproject/sof/commits/main/.github is enough.

So will enabling this particular parameter name check in CI help detect some real problems? I can't imagine any for the moment but I'd love to be proven wrong.


>  because the compiler knows what to do.

I think there is a big difference between code and comments ignored by the compiler.  Both code and comments should be good (see the many  doxygen and comment fixes I made in #4509, #4441,  #4106, #3912, #2626,...) but there's a clear line between code and comments. Parameter names in declarations are not code, they're (important!) comments.

Comment (plbossart) [reply src/include/sof/ipc/common.h:190]: cppcheck is still to verbose to be used for CI, specifically it does not work well with unions and our friends at AMD used unions. Coverity is the same, too verbose.

It's more of a maintainer tool that is run on a e.g. weekly basis - and before releases as we didn't do for 1.9... Oh well.

**Comment by lgirdwood** [comment]
@cujomalainey are you able to align the doxygen comment? We are good to go once this is done. Thanks !

File: src/drivers/intel/dmic/dmic_nhlt.c
Comment (lyakh) [inline src/drivers/intel/dmic/dmic_nhlt.c:192]: `ref` can be uninitialised below if neither `DMIC_IPM_VER1` nor `DMIC_IPM_VER2` is defined. Probably this cannot happen, so it isn't really important what happens then, but in case there is such a possibility, with this initialisation we'd probably hit an error:
```
		if (ref != val) {
			dai_err(dai, "dmic_set_config_nhlt(): illegal OUTCONTROL%d = 0x%08x",
				n, val);
			return -EINVAL;
		}

```
Again - probably unimportant, but just to be aware of.

File: src/drivers/intel/dmic/dmic_computed.c
Comment (lyakh) [inline src/drivers/intel/dmic/dmic_computed.c:135]: this is an actual flow change, is this intended? @singalsu is this correct? If this is fixing an error, we should indicate that. Also not sure why you're saying that this check is pointless?

File: src/include/sof/ipc/common.h
Comment (lyakh) [reply src/include/sof/ipc/common.h:190]: I've seen the same warning - mismatched parameter name in declaration and definition without cppcheck. I think the compiler checks it itself (maybe enabled by some flags, or added in some versions), or maybe it was one of the tools, used during the kernel / SOF / Zephyr build. I wasn't using anything special for sure.

File: src/drivers/intel/dmic/dmic_computed.c
Comment (singalsu) [reply src/drivers/intel/dmic/dmic_computed.c:135]: Yep, there's a mistake, also a duplicate decimation factor in fir_list[1] should be checked for. The check should have been `if (j > 0 &&` but check for non-zero is fine too. The commit title and description could be changed.

**Comment by lyakh** [comment]
> @cujomalainey I guess you are busy so @lyakh will fix it up for you.

#4864

**Comment by cujomalainey** [comment]
Sounds good, will reopen when i have time to address the comments

**Comment by cujomalainey** [comment]
Thanks marks, I forgot my chain was apart of that

And you can do prs against each other's repos to land stuff on remote branches

**Comment by marc-hb** [comment]
> And you can do prs against each other's repos to land stuff on remote branches

Yes but you cannot change someone else's fork so a new PR was required in this case.

**Pull request closed without merging**
